File: day1/day1.erl
-module(day1).
-export([part1/0, part2/0]).
-define(FILE_NAME, "input.txt").


part1() -> count(read_input()).

read_input() ->
    case file:read_file(?FILE_NAME) of
	{ok, Bin} -> lists:map(
		       fun(X) -> {Int, _} = string:to_integer(X), Int end,
		       string:lexemes(Bin, "\n")
		      );
	{error, _} -> throw("Unable to open file")
    end.


c2([_], C) -> C;
c2([A,B | T], C) ->
    if
		B > A -> c2([B|T], C+1);
		true -> c2([B|T], C)
    end.

count([]) -> 0;
count([_]) -> 0;
count([H|L]) -> {_, Final} = 
		lists:foldl(
		  fun(X, {Prev, Count}) when X > Prev -> {X, Count+1};
		     (X, {_Prev, Count}) -> {X, Count} end,
		  {H, 0},
		  L), Final.


part2() -> count( three_mesurment(read_input(), [])).

three_mesurment([F,S,T | Other], Acc) -> three_mesurment([S,T | Other], [F+S+T | Acc]);
three_mesurment([_,_], Acc) -> lists:reverse(Acc).

File: day1/Day1.ex
defmodule Day1 do
@filename "input.txt"

def read_input() do
    case :file.read_file(@filename) do
        {:ok, bin} -> String.split(bin, "\n")
                        |> Stream.filter(fn elem -> elem != "" end)
                        |> Stream.map(&:string.to_integer/1)
                        |> Stream.map(fn {fst, _} -> fst end)
                        |> Enum.to_list
        {:error, why} -> throw(["Unable to open file", why])
    end
end

def count([]), do: 0
def count([_]), do: 0
def count([h|t]) do
  List.foldl(
    t,
    {h, 0},
    fn(elem, {prev, count}) when elem > prev -> {elem, count+1};
      (elem, {_prev, count}) -> {elem, count}
    end)
    |> Kernel.elem(1) # Attention, zero based indexing
end

def part1(), do: read_input() |> count

def three_mes([_,_], acc), do: Enum.reverse(acc)
def three_mes([a,b,c | tail], acc) do
  three_mes([b, c | tail], [a+b+c | acc])
end

def part2(), do: read_input() |> three_mes([]) |> count

end
